core/dnsserver/watch.go

```diff
@@ -0,0 +1,153 @@
+package dnsserver
+
+import (
+	"io"
+	"log"
+	"sync"
+
+	"github.com/miekg/dns"
+
+	"github.com/coredns/coredns/pb"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/watch"
+)
+
+// watcher contains all the data needed to manage watches
+type watcher struct {
+	changes watch.Chan
+	counter int64
+	watches map[string]watchlist
+	plugins []watch.Watchable
+	mutex   sync.Mutex
+}
+
+type watchlist map[int64]pb.DnsService_WatchServer
+
+func newWatcher(zones map[string]*Config) *watcher {
+	w := &watcher{changes: make(watch.Chan), watches: make(map[string]watchlist)}
+
+	for _, config := range zones {
+		plugins := config.Handlers()
+		for _, p := range plugins {
+			if x, ok := p.(watch.Watchable); ok {
+				x.SetWatchChan(w.changes)
+				w.plugins = append(w.plugins, x)
+			}
+		}
+	}
+
+	//TODO: maybe a stop channel, work properly with reloads?
```

Comment:
I think we need a stop channel to handle reloads properly.

Reply:
Sorry - can you explain more?  I understand that i need to do something to make the reload plugin work here... I don't understand specifically what I need to do with a stop channel.  Is there an example of something else that uses a stop channel in relation to the reload plugin?

Reply:
Look at `plugin/pkg/up/up.go`. Or `grep stop -r plugin` and you'll see some more examples. Basically you have `processWatches` do its `select` on both channels, and if it gets anything on the stop channel, it exits (stops the go routine).

Then in OnShutdown you send `true` down the stop channel. We might also want to notify all the clients that we are stopping the watches at that time, too. And close the streams maybe? Clean up at least.